src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,171 @@
+import RequestQueue from "./RequestQueue";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Resolver = (value?: any) => void;
+type Rejecter = (reason?: unknown) => void;
+
+type RequestSubscription = {
+  subscriberId: number;
+  resolve: Resolver;
+  reject: Rejecter;
+};
+
+/**
+ * An extension of `RequestQueue` that adds a concept of "subscribers," which may share references to a single request
+ * or cancel their subscription without disrupting the request for other subscribers.
+ */
+export default class SubscribableRequestQueue {
+  private queue: RequestQueue;
+
+  /** Number of subscribers, used for generating unique subscriber IDs. */
+  private subscriberIds: number;
+  /** Map keyed by subscriber ID. Subscribers are only useful for cancelling early, so we only store rejecters here. */
+  private subscribers: Map<number, Map<string, Rejecter>>;
+  /** Map from "inner" request (managed by `queue`) to "outer" promises generated per-subscriber. */
+  private requests: Map<string, RequestSubscription[]>;
+
+  constructor(maxActiveRequests?: number) {
+    this.queue = new RequestQueue(maxActiveRequests);
+    this.subscriberIds = 0;
+    this.subscribers = new Map();
+    this.requests = new Map();
+  }
+
+  /** Resolves all subscriptions to request `key` with `value` */
+  private resolveAll<T>(key: string, value: T) {
```

Comment:
I'm surprised the linter doesn't require you to add explicit void return types on these. Maybe the Nucmorph-Colorizer project has different settings.